src/internal/storage/fileset/validator.go

```diff
@@ -0,0 +1,66 @@
+package fileset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
+)
+
+var _ FileSet = &pathValidator{}
+
+type pathValidator struct {
+	fs FileSet
+}
+
+// NewPathValidator checks if there is collision between file and dir names.
```

Comment:
This abstraction checks for more than just file and directory path collisions, so I think we should just say `NewPathValidator validates the paths in a file set` for now.

Reply:
Yea, I'll change that.